### PR TITLE
CCI14 add USSD post op registration flow

### PR DESCRIPTION
**Pull request by NicolaasDieBaas**
(https://praekeltorg.atlassian.net/browse/CCI-14)
- USSD flows (https://drive.google.com/open?id=0B1uSxTPWAbJRU2JVSVBWejlXaWM)

### CODE REVIEW

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps please review

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps please review

File: src/ussd_app.js [hunk 1]
```diff
@@ -126,7 +126,145 @@ go.app = function() {
...
+                choices: [
+                    new Choice("state_consent", $("Today")),
+                    new Choice("state_consent", $("Yesterday")),
+                    month_choice[0],
+                    month_choice[1],
+                    month_choice[2],
+                    new Choice("state_pre_op", $("I haven't had my operation yet"))
+                ],
+                next: function(choice) {
+                    //self.im.user.set_answer("op_date", )  // save month and year first, then add day at next state
```
Comment (imsickofmaps) [inline src/ussd_app.js]: Remove or activate?

File: src/ussd_app.js [hunk 2]
```diff
@@ -126,7 +126,145 @@ go.app = function() {
...
+            return new FreeText(name, {
+                question: $([
+                    "Please input the day you had your operation. For example, "
+                    + "12.",
+                ].join("")),
+                next: function(text) {
+                    // add a zero to input if a single-digit number
+                    if (text.length == 1) text = "0" + text;
+
+                    if (go.utils.is_date_diff_less_than_6weeks(self.im, month_year+text)) {
```
Comment (imsickofmaps) [inline src/ussd_app.js]: can we make this util a bit more utilly? As in make the "6 weeks" an option...

File: src/ussd_app.js [hunk 3]
```diff
@@ -126,7 +126,145 @@ go.app = function() {
...
+                    "holidays, weekends & before 8am?\n",
+                    "- Having ur cell# & language info stored so we can send u",
+                    " SMSs?"
+                ].join("")),
+                choices: [
+                    new Choice("yes", $("Yes")),
+                    new Choice("state_consent_withheld", $("No"))
+                ],
+                next: function(choice) {
+                    return choice.value;
```
Comment (imsickofmaps) [inline src/ussd_app.js:196]: if they select "yes" where does this route?

File: src/ussd_app.js [hunk 4]
```diff
@@ -126,7 +126,145 @@ go.app = function() {
...
+            return new ChoiceState(name, {
+                question: $([
+                    "Do you consent to:\n",
+                    "- Receiving some SMSs on public ",
+                    "holidays, weekends & before 8am?\n",
+                    "- Having ur cell# & language info stored so we can send u",
+                    " SMSs?"
+                ].join("")),
+                choices: [
+                    new Choice("yes", $("Yes")),
```
Comment (imsickofmaps) [inline src/ussd_app.js]: these are not hugely consistent

File: src/ussd_app.js [hunk 5]
```diff
@@ -126,7 +126,145 @@ go.app = function() {
             });
         });
 
-        self.states.add('state_servicerating_location', function(name){
+        // ChoiceState st-F1
+        self.states.add('state_op', function(name) {
+            var today = go.utils.get_today(self.im.config);
+            var month_choice = go.utils.make_month_choices($, today, 3, 1, "YYYYMM", "MMMM 'YY");
```
Comment (imsickofmaps) [inline src/ussd_app.js]: switch to backwards `-1`

File: src/ussd_app.js [hunk 6]
```diff
@@ -254,12 +392,15 @@ go.app = function() {
...
+                    return {
+                        name: choice.value,
+                        creator_opts: false
+                    };
                 }
             });
         });
 
-        self.states.add('state_bfl_join', function(name){
+        self.states.add('state_bfl_join', function(name, post_op_registration) {
```
Comment (imsickofmaps) [inline src/ussd_app.js]: Break these out for sanity

File: src/ussd_app.js [hunk 7]
```diff
@@ -271,12 +412,22 @@ go.app = function() {
...
+                    } else {
+                        return choice.value;
+                    }
+
                 }
             });
         });
 
-        self.states.add('state_bfl_no_join', function(name){
+        self.states.add('state_bfl_no_join', function(name, post_op_registration) {
```
Comment (imsickofmaps) [inline src/ussd_app.js]: And this

File: test/ussd_app.test.js [hunk 8]
```diff
@@ -141,6 +143,178 @@ describe("app", function() {
...
+
+        describe("when user select a date that is < 6 weeks", function() {
+            it("should ask for consent", function() {
+                return tester
+                    .setup.user.state('state_main_menu')
+                    .inputs('3', '4', '5')
+                    .check.interaction({
+                        state: 'state_consent',
+                        reply: [
+                            "Do you consent to:\n"
```
Comment (imsickofmaps) [inline test/ussd_app.test.js]: lets be consistent in the usage of array to add the newlines

**Comment by imsickofmaps** [comment]
@NicolaasDieBaas first pass done

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps another review pass please sir

**Pull request merged**